text/0000-undisambiguated-generics.md

```diff
@@ -0,0 +1,236 @@
+the feature is enabled. This is undesirable in most existing codebases, as the number of
+linted expressions is likely to be large, but could be useful for new codebases and in the future.
+
+Note that, apart from for those users who explicitly increase the level of the lint, no steps are
+taken to discourage the use of `::` at this stage (including in tools, such as rustfmt).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The primary drawback is that resolving ambiguities in favour of generics means changing the
+interpretation of `(a<b, c>(d))` from a pair of tuples to a generic function call. However this has
+been demonstrated ([1](https://github.com/rust-lang/rust/pull/53578#issuecomment-421475443)) not to
+cause issues in practice (the syntax is unnatural for Rust and is actively warned against by the
+compiler).
+
+Additionally, there is potential for performance regressions due to backtracking. However,
+empirical evidence ([1](https://github.com/rust-lang/rust/pull/53511#issuecomment-414172984) and
+[2](https://github.com/rust-lang/rust/pull/53511#issuecomment-414360849)) suggests this should not
+be a problem. Although it is probable that a pathological example could be constructed that does
+result in poorer performance, such an example would not be representative of typical Rust code and
+therefore is not helpful to seriously consider. Backtracking is already used for some cases in the
+parser.
+
+The other potential drawback is that other parsers for Rust's syntax (for example in external tools)
+would also have to implement some form of backtracking (or similar) to handle this case. However,
+backtracking is straightforward to implement in many forms of parser (such as recursive decent) and
+it is likely this will not cause significant problems.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+If we want to allow `::` to be omitted, there are two solutions:
+- Backtracking, as suggested here.
+- Using a parser for nondeterministic grammars, such as GLL.
+
+Although using a more sophisticated parser would come with its own advantages, it's an overly
+complex solution to this particular problem. Backtracking seems to work well in typical codebases
+and provides an immediate solution to the problem.
+
+Alternatively we could continue to require `::`. This would ensure there would be no performance
+implications, but would leave the nonconformal and surprising syntax in place. We could potentially
+use backtracking to provide the improved diagnostic suggestions to use `::`, while still preventing
+`::` from being omitted.
+
+## Future frequency of disambiguated generic expressions
+It is likely that, should the
+[generalised type ascription](https://github.com/rust-lang/rfcs/pull/2522) RFC be accepted and
+implemented, the number of cases where generic type arguments have to be provided is reduced, making
+users less likely to encounter the `::` construction. However, the
```

Comment:
In particular you can note that type ascription can be less ergonomic when the quantified type parameter occurs applied to nested type constructors, e.g:

```rust
fn foo<T>() -> Vec<HashSet<T>> { ... }

// ascription:

let x = foo() : Vec<HashSet<u8>>;

// type application:

let x = foo<u8>();
```